### ISSUE TITLE
Create an editor to edit images before uploading

### ISSUE THREAD

**Issue by miketaylr** [webcompat/webcompat.com#1658]
(I can't remember if there's already an issue for this... searched around and found nothing)

It would be good to have the ability to do the following to an image, before uploading:

1. highlight in yellow (paintbrush or rectangle tool)
2. black out an area (paintbrush or rectangle tool)
3. blur an area (rectangle)

This is all doable with modern canvas API support, but the blurring will need a polyfill written (possibly running in a web worker) for Safari and others.

(update) tasks breakdown

- [x] update demo (current canvas code stub) so lines are drawn on drag event (before the mouse up which draws entire rectangle)
- [x] decide where image editor code should go (bugform.js or new file); going with bugform.js for testing ease for now (but may change in future); in case it gets unruly adding in comments like // image editor props to be able to move stuff to another file
- [x] add logic to create a popup/modal (image editor)... need Edit btn (click handler and tracking Edit state)
- [x] add Editor control buttons (remove cur placeholders outside of canvas and create html/css buttons); hook buttons up to existing click event handlers
- [x] integrate cur demo code (canvas stub) into webcompat code
- [x] get screenshot img to show in canvas background
- [x] add Exit/close control button (update state to non Edit mode); hook up handler
- [x] update logic to save final canvas image on form post
- [x] fix screenshot and canvas resizing (half working, when I can get the original image dimensions, the calculations are working but currently having a challenge with getting small image dimensions, ie screenshots or non-complex photos)
- [x] follow up screenshot/canvas resizing; couple of cases: 1 orig screenshot width/height somehow unavailable, then a default canvas width/height should be set and img doesn't need resizing; 2 orig screenshot is much smaller than the canvas width; then screenshot doesn't need resizing
- [x] fix controls (figure out better buttons/design); 
* X- add css for hover/active styles (use cur submit button styles as reference)
* X- toolbar buttons for Image Editor (Highlight btn, Redact btn); 
* X- Form File Input > filled state (Edit btn); 
* X- need apply flexbox to canvas and toolbar area (in place of float right)
- [x] shape drawing (black rectangle) on drag
- [ ] fix line drawing (yellow rectangle) on drag; currently working but not perfect
- [x] internal issues (bugs tracked in the prd currently)
* X - handling ending mouseup position on a smaller shape size than originally drawn during drag motions
* X - canvas placement looks different on FF vs Chrome
- [ ] fine tune layout; general Editor layout (where tools go, how canvas lies); the canvas alignment to modal is off by 1-2 px here and there (probably sizing); probably need to adjust height of the container b/c of the rounded corners or perhaps do away with rounded corner?
- [ ] functional tests
- [ ] start thinking about mobile support

**Comment by cch5ng** [comment]
discussion: blur is a nice to have feature

Tom: blur could be a security issue

**Comment by cch5ng** [comment]
@miketaylr questions:
* at what point in the flow do you expect image editing to happen?
  * only before the initial issue is saved?
  * also after the issue has been created? (after issue has been saved, user can upload additional images, do these need an editor feature?)

* is canvas the only/best option? is it possible to use svg for this? would there be any benefits?

* do you need to keep a copy of the original image? (with edits as an overlay) or is the expectation to save annotations as changes to the original image (saving only one file)

* how frequently should the image get saved (ie before the file upload)?
  * if it should get saved before file upload 
    * where should it get saved (before the file upload happens on form submit)? in browser storage?

* do you need to support things like undo/redo (how many steps)

* do you want the image to be editable in a separate tab/window/popup (the default image size looks a bit small)

* (proposal) is there any flexibility in requirements to extend features to include choosing colors (other than yellow/black)? (maybe a color picker something similar to choosing label colors in github)
  * arguments for
    * yellow might not always provide enough contrast against the website background/section
    * gives people a sense of choice/freedom
    * potentially makes people happy (if reporters are generally super techy maybe this feature is more neutral)
  * arguments against
    * potentially adds another couple steps to the user flow

* is it expected to extend the features of the editor in the future or do you expect it would only ever have the 2-3 current features

**Comment by miketaylr** [comment]
> at what point in the flow do you expect image editing to happen?
> 
> only before the initial issue is saved?

For now, let's scope it to this. The user should be able to edit their screenshot/image upload from the New Issue form, before upload.

> also after the issue has been created? (after issue has been saved, user can upload additional images, do these need an editor feature?)

Seems like a good follow up issue.

**Comment by miketaylr** [comment]
> is canvas the only/best option? is it possible to use svg for this? would there be any benefits?

I don't know why we would use svg for editing bitmaps/manipulating pixels, so let's stick to canvas.

**Comment by miketaylr** [comment]
> do you need to keep a copy of the original image? (with edits as an overlay) or is the expectation to save annotations as changes to the original image (saving only one file)

we don't want to keep the original -- the point of the feature is for the user to be able to edit sensitive information. if we keep the original... there's no real point. we might consider creating an undo/redo stack of edits, but that should be follow-up work.

**Comment by miketaylr** [comment]
> how frequently should the image get saved (ie before the file upload)?

Only once, when it's POSTed to the server.

**Comment by miketaylr** [comment]
> do you need to support things like undo/redo (how many steps)

Not for this first version. Let's file a follow up issue after we ship v1.

**Comment by miketaylr** [comment]
> do you want the image to be editable in a separate tab/window/popup (the default image size looks a bit small)

Possibly? Some kind of modal-ish dialog could work. Keep in mind we'll want this to work for mobile users at some point as well (even if not v1).

**Comment by miketaylr** [comment]
> (proposal) is there any flexibility in requirements to extend features to include choosing colors (other than yellow/black)? (maybe a color picker something similar to choosing label colors in github)

I would prefer for this to be follow up work. v2 or beyond.

**Comment by miketaylr** [comment]
> is it expected to extend the features of the editor in the future or do you expect it would only ever have the 2-3 current features

For now, I'd like it to be simple. Things can change in the future, but let's plan for the restricted feature set.

**Comment by miketaylr** [comment]
> blur an area (rectangle)

Note: let's skip blur.

**Comment by miketaylr** [comment]
> any other questions that should be here?

Just thinking through how we might make this work for touch/mobile and desktop users. v1 should be restricted to desktop, but ideally we have mobile support at some point after.

**Comment by miketaylr** [comment]
> is canvas the only/best option? is it possible to use svg for this? would there be any benefits?
> 
>> I don't know why we would use svg for editing bitmaps/manipulating pixels, so let's stick to canvas.

To expand here, here's what I imagine we do to get started:

1. someone clicks edit (or something)
1. the user is presented with editing tools
1. a canvas element is created, the existing dataURI preview image is `drawImage`'d into the canvas element
1. the user manipulates the canvas pixels with the tools, via canvas APIs
1. when the user clicks save (or something), the canvas is `toDataURL`'d and the resulting data becomes the new preview (and will be what gets uploaded).

**Comment by cch5ng** [comment]
@miketaylr there's a quick demo just for the rectangle drawing parts (your 4):
https://cch5ng.github.io/image-edit/

(the yellow square is a placeholder for the highlight rectangle button; the blue square is a placeholder for the redact rectangle button)

it's rough and I'll change details like the controls will be moved out of the canvas and have a smaller graphic button. other steps are to integrate the logic with webcompat.com and handle your steps:
1, 2 (partly there but needs details and cleanup), 3, 5

**Comment by cch5ng** [comment]
hi @miketaylr, I'm pretty sure my initial thoughts for handling the mouse drag (to draw the rectangle borders in real time) were not complete and am trying to rethink about the logic. could I get a sanity check if my current logic seems reasonable? 

I feel like I've done something similar in the past where I was doing a constant diff between an old mouse position (from previous point in time) to the current mouse position... do you think the line drawing needs to be that granular?

points of info need to track
* original mouse position on initial click (mouse down)
* whether the user starts initial drag motion in x or y axis (probably need to set a threshold for error for wobbly mouse drag)
* when the user shifts the axis of their final drag motion (to close the rect), need to get the current points for ends of each line (line 1 and line 2 starting from original mouse position => become the start points for lines 3 and 4); also get the cur axis of movement
* cur mouse position on mouse up event (as the final point for lines 3 and 4)

**Comment by miketaylr** [comment]
> when the user shifts the axis of their final drag motion (to close the rect), need to get the current points for ends of each line (line 1 and line 2 => become the start points for lines 3 and 4); also get the cur axis of movement

What do you mean by shift axis? Are line 1 and line 2 left and top, respectively?

> (if you're too busy, I will probably start working along these lines but thought maybe I should get a double check in case this is over thinking/complicating...)

That sounds reasonable to me, if I'm understanding what you're saying correctly.

**Comment by miketaylr** [comment]
Closing this, will open a follow up later.
